server/src/main/java/org/elasticsearch/search/lookup/SourceLookup.java

```diff
@@ -104,19 +107,23 @@ public int docId() {
     }
 
     public void setSegmentAndDocument(LeafReaderContext context, int docId) {
+        // if we are called with the same document, don't invalidate source
         if (this.reader == context.reader() && this.docId == docId) {
-            // if we are called with the same document, don't invalidate source
             return;
         }
+
+        // only reset reader and fieldReader when reader changes
         if (this.reader != context.reader()) {
             this.reader = context.reader();
-            // only reset reader and fieldReader when reader changes
-            if (context.reader()instanceof SequentialStoredFieldsLeafReader lf) {
-                // All the docs to fetch are adjacent but Lucene stored fields are optimized
-                // for random access and don't optimize for sequential access - except for merging.
-                // So we do a little hack here and pretend we're going to do merges in order to
-                // get better sequential access.
-                fieldReader = lf.getSequentialStoredFieldsReader()::visitDocument;
+
+            // All the docs to fetch are adjacent but Lucene stored fields are optimized
+            // for random access and don't optimize for sequential access - except for merging.
+            // So we do a little hack here and pretend we're going to do merges in order to
+            // get better sequential access.
+            if (context.reader() instanceof SequentialStoredFieldsLeafReader lf) {
+                // Avoid eagerly loading the stored fields reader, since this can be expensive
+                Supplier<StoredFieldsReader> supplier = new MemoizedSupplier<>(lf::getSequentialStoredFieldsReader);
```

Comment:
The main issue is that `getSequentialStoredFieldsReader` calls `StoredFieldsReader#getMergeInstance`, which creates a new stored fields reader. I wonder if this could be optimized in Lucene too to avoid recreating it every time 🤔